scripts/generate/templates/Heap.js

```diff
@@ -0,0 +1,322 @@
+}
+
+struct ${node} {
+    ${valueType} value;
+    ${indexType} index; // position -> value
+    ${indexType} lookup; // value -> position
+}
+
+/**
+ * @dev Lookup the root element of the heap.
+ */
+function peek(${struct} storage self) internal view returns (${valueType}) {
+    // self.data[0] will \`ARRAY_ACCESS_OUT_OF_BOUNDS\` panic if heap is empty.
+    return _unsafeNodeAccess(self, self.data[0].index).value;
+}
+
+/**
+ * @dev Remove (and return) the root element for the heap using the default comparator.
+ *
+ * Note: All inserting and removal from a heap should always be done using the same comparator. Mixing comparator
+ * during the lifecycle of a heap will result in undefined behavior.
+ */
+function pop(${struct} storage self) internal returns (${valueType}) {
+    return pop(self, Comparators.lt);
+}
+
+/**
+ * @dev Remove (and return) the root element for the heap using the provided comparator.
+ *
+ * Note: All inserting and removal from a heap should always be done using the same comparator. Mixing comparator
+ * during the lifecycle of a heap will result in undefined behavior.
+ */
+function pop(
+    ${struct} storage self,
+    function(uint256, uint256) view returns (bool) comp
+) internal returns (${valueType}) {
+    unchecked {
+        ${indexType} size = length(self);
+        if (size == 0) Panic.panic(Panic.EMPTY_ARRAY_POP);
+
+        ${indexType} last = size - 1;
+
+        // get root location (in the data array) and value
+        ${node} storage rootNode = _unsafeNodeAccess(self, 0);
+        ${indexType} rootIdx = rootNode.index;
+        ${node} storage rootData = _unsafeNodeAccess(self, rootIdx);
+        ${node} storage lastNode = _unsafeNodeAccess(self, last);
+        ${valueType} rootDataValue = rootData.value;
+
+        // if root is not the last element of the data array (that will get pop-ed), reorder the data array.
+        if (rootIdx != last) {
+            // get details about the value stored in the last element of the array (that will get pop-ed)
+            ${indexType} lastDataIdx = lastNode.lookup;
+            ${valueType} lastDataValue = lastNode.value;
+            // copy these values to the location of the root (that is safe, and that we no longer use)
+            rootData.value = lastDataValue;
+            rootData.lookup = lastDataIdx;
+            // update the tree node that used to point to that last element (value now located where the root was)
+            _unsafeNodeAccess(self, lastDataIdx).index = rootIdx;
+        }
+
+        // get last leaf location (in the data array) and value
+        ${indexType} lastIdx = lastNode.index;
+        ${valueType} lastValue = _unsafeNodeAccess(self, lastIdx).value;
+
+        // move the last leaf to the root, pop last leaf ...
+        rootNode.index = lastIdx;
+        _unsafeNodeAccess(self, lastIdx).lookup = 0;
+        self.data.pop();
+
+        // ... and heapify
+        _siftDown(self, last, 0, lastValue, comp);
```

Comment:
Wouldn't this be needed only if `rootIdx !== last` as well? Otherwise my understanding is that the order is preserved so it may be unnecessary

Reply:
if `rootIdx !== last`, what we did is move the root value to the end of the array, so it can safelly be popped. we shifted around the `value` part of the array, and made sure the indices are still correct. 

For that last part we made sure that the index that used to point to the last value now point to the location that the root used to point to (that stored the data that we pop).

This shifting in the values do NOT change the tree. Each node of the tree still points to the same value (but possibly at a different location).

Just before pop-ing the array we did `rootNode.index = lastIdx;`, which is basically saying 
> we take the last leaf, which could be anything, put it where the root was, and then propagate it down. 

There is no reason to think that this last root is already properly placed. As far as we know it could be the biggest value in the heap.

We have fuzzing tests, so fell free to give it a try. I don't think your logic is good, but maybe I'm missing something.